### ISSUE TITLE
KeyError: 'getpwuid(): uid not found: 472'

### ISSUE THREAD

**Issue by mattcarrier** [OmegaVVeapon/kopf-k8s-sidecar#12]
In order to resolve the WATCH hang issue with the grafana helm chart I switched over to this sidecar from kiwigrid.  Whenever the sidecar starts up I keep getting uid not found for 472:

```
# kubectl logs isa-kube-prometheus-grafana-b6669d485-v6wbz -c grafana-sc-datasources
Traceback (most recent call last):
  File "/.venv/bin/kopf", line 8, in <module>
[2021-01-20 17:31:42,887] kopf.activities.star [INFO    ] Folder /etc/grafana/provisioning/datasources already exists. Skipping creation.
[2021-01-20 17:31:42,887] kopf.activities.star [INFO    ] Client watching requests using a timeout of 660 seconds
[2021-01-20 17:31:42,887] kopf.activities.star [INFO    ] Server watching requests using a timeout of 600 seconds
    sys.exit(main())
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/kopf/cli.py", line 50, in wrapper
    return fn(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/kopf/cli.py", line 101, in run
    vault=__controls.vault,
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 107, in run
    vault=vault,
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 149, in operator
    vault=vault,
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 249, in spawn_tasks
    id=peering.detect_own_id(), priority=priority,
  File "/.venv/lib/python3.7/site-packages/kopf/engines/peering.py", line 300, in detect_own_id
    user = getpass.getuser()
  File "/usr/local/lib/python3.7/getpass.py", line 169, in getuser
    return pwd.getpwuid(os.getuid())[0]
KeyError: 'getpwuid(): uid not found: 472'
```

The comment that pointed me to this sidecar:
https://github.com/kiwigrid/k8s-sidecar/issues/85#issuecomment-749681385

**Comment by OmegaVVeapon** [comment]
Interesting...
Could you provide more info on how you're getting this? 

Are you using this image as a base and adding on top of it? If so, provide the Dockerfile.
If not, could it be that you're running it with a k8s `SecurityContext` that introduces that `uid` of `472`?

Bear in mind that I'm using [uid/gid `65534`](https://github.com/OmegaVVeapon/kopf-k8s-sidecar/blob/main/Dockerfile#L39) in the same way that the [original k8s-sidecar does](https://github.com/kiwigrid/k8s-sidecar/blob/master/Dockerfile#L15) so you shouldn't have uid issues in one but not the other...

```
❯ docker run --rm -it omegavveapon/kopf-k8s-sidecar sh
$ whoami
nobody
$ id
uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup)
```

**Comment by mattcarrier** [comment]
> Interesting...
> Could you provide more info on how you're getting this?
 
Yes, absolutely.  I'm using [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack) which uses the grafana chart as a dependency.  In the values I swapped out the kiwigrid sidecar for this sidecar.

> Are you using this image as a base and adding on top of it? If so, provide the Dockerfile.

Nope, using the out of the box image. All I did was swap out the sidecar repository and image in the kube-prometheus-stack values.yaml.

```
    sidecar:

      # Connection hanging bug in kiwigrid/k8s-sidecar:
      # https://github.com/kiwigrid/k8s-sidecar/issues/85
      #
      # omegavveapon/kopf-k8s-sidecar is supposed to solve this issue but when I replaced it I got the following error:
      # https://github.com/OmegaVVeapon/kopf-k8s-sidecar/issues/12
      #
      # Going back to kiwigrid/k8s-sidecar in the interim until we hear back from OmegaVVeapon. This means that you may
      # have to restart the sidecar if you don't see your dashboards being added/updated/deleted correctly.

      image:
        repository: kiwigrid/k8s-sidecar
        tag: 0.1.151
        #repository: omegavveapon/kopf-k8s-sidecar
        #tag: 1.1.0
```

> If not, could it be that you're running it with a k8s `SecurityContext` that introduces that `uid` of `472`?

So something I did come across is this:

https://grafana.com/docs/grafana/latest/installation/docker/#migrate-to-v51-or-later (this is the migration of UID to 472)

We are using 7.1.1, are you using a version prior to 5.1?

> 
> Bear in mind that I'm using [uid/gid `65534`](https://github.com/OmegaVVeapon/kopf-k8s-sidecar/blob/main/Dockerfile#L39) in the same way that the [original k8s-sidecar does](https://github.com/kiwigrid/k8s-sidecar/blob/master/Dockerfile#L15) so you shouldn't have uid issues in one but not the other...
> 
> ```
> ❯ docker run --rm -it omegavveapon/kopf-k8s-sidecar sh
> $ whoami
> nobody
> $ id
> uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup)
> ```

Yes, this confuses me greatly as well, but indeed kiwigrid works.

**Comment by OmegaVVeapon** [comment]
Thanks for the info, very helpful!

So I can repro the issue you're seeing.

```
❯ docker run --env FOLDER=test --env LABEL=asds --rm --user 472 -it omegavveapon/kopf-k8s-sidecar
[2021-01-21 15:59:00,080] kopf.activities.star [INFO    ] Creating folder test
Traceback (most recent call last):
  File "/.venv/bin/kopf", line 8, in <module>
[2021-01-21 15:59:00,080] kopf.activities.star [ERROR   ] Insufficient privileges to create folder test.
[2021-01-21 15:59:00,080] kopf.activities.star [INFO    ] Client watching requests using a timeout of 660 seconds
[2021-01-21 15:59:00,080] kopf.activities.star [INFO    ] Server watching requests using a timeout of 600 seconds
    sys.exit(main())
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/kopf/cli.py", line 50, in wrapper
    return fn(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/.venv/lib/python3.7/site-packages/kopf/cli.py", line 101, in run
    vault=__controls.vault,
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 107, in run
    vault=vault,
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 149, in operator
    vault=vault,
  File "/.venv/lib/python3.7/site-packages/kopf/reactor/running.py", line 249, in spawn_tasks
    id=peering.detect_own_id(), priority=priority,
  File "/.venv/lib/python3.7/site-packages/kopf/engines/peering.py", line 300, in detect_own_id
    user = getpass.getuser()
  File "/usr/local/lib/python3.7/getpass.py", line 169, in getuser
    return pwd.getpwuid(os.getuid())[0]
KeyError: 'getpwuid(): uid not found: 472'
```

But to me... that's perfectly normal?
You can't make a process run under a uid that doesn't exist in Unix... I have no idea how the original k8s-sidecar is working. 

Will have to investigate that.

That being said, while I respect that grafana wants to run under user `472`, it doesn't make sense that the sidecar would also run under that `uid`. 

[Grafana creates uid 472](https://github.com/grafana/grafana/blob/master/Dockerfile#L39) inside the image in the same way that we create uid `65534`. It makes no sense to run both of them using `65534` **OR** `472`... 

I could "fix" this by simply switching the uid to match Grafana's `472` uid but then, what about the next user that asks to run the sidecar for uid `777` or something else...?

I'll have to have a think about this

**Comment by djsly** [comment]
isn't this helping with access on the files, since a emptyDir is shared between both containers ?

**Comment by mattcarrier** [comment]
> isn't this helping with access on the files, since a emptyDir is shared between both containers ?

Sorry @djsly I'm not sure I understand your question.  But yes, Grafana utilizes the sidecar to synchronize dashboards.

**Comment by schmidt-i** [comment]
We have the same issue. We tried to use the kopf-k8s-sidecar as an alternative to the kiwigrid one a prometheus operator setup. I saw in the codebase of kopf, that this error should be gone if we set the environment variable POD_ID to a value. However this is not possible with the prometheus-operator helmchart, since we are not able to adjust the sidecars environment variables. Would it be possible for you to set the POD_ID environment variable in the Dockerfile to any useful value? See this code lines for more information: https://github.com/nolar/kopf/blob/master/kopf/engines/peering.py#L245

**Comment by OmegaVVeapon** [comment]
> We have the same issue. We tried to use the kopf-k8s-sidecar as an alternative to the kiwigrid one a prometheus operator setup. I saw in the codebase of kopf, that this error should be gone if we set the environment variable POD_ID to a value. However this is not possible with the prometheus-operator helmchart, since we are not able to adjust the sidecars environment variables. Would it be possible for you to set the POD_ID environment variable in the Dockerfile to any useful value? See this code lines for more information: https://github.com/nolar/kopf/blob/master/kopf/engines/peering.py#L245

Thanks for that find, would it possible? Certainly.

Just curious, have you tried to build the image with that env var on your side to see if that resolves the issue?

Another possible solution would be to set the `USER` env var, according to the source Python code
https://github.com/python/cpython/blob/v3.6.0/Lib/getpass.py#L155-L170

I gave it quick try with: 
`docker run --env USER=412 --env FOLDER=test --env LABEL=asds --rm --user 472 -it omegavveapon/kopf-k8s-sidecar` 

and it no longer showed the error, need to try it on a real k8s `Pod` though to be 100% sure... I'll be doing that today.

**Comment by OmegaVVeapon** [comment]
@mattcarrier @schmidt-i 
Hey guys, I built a dev release with a proposed fix and have tested it using multiple different `uid`s using k8s `SecurityContext`s and everything seems to working fine.

You can get the release from `omegavveapon/kopf-k8s-sidecar:fix-uid-issues-with-grafana-chart`.

If everything is working fine for you as well, I'll create a tag from it.

**Comment by mattcarrier** [comment]
@OmegaVVeapon Thanks!  I'll give it a go ASAP.  Expect to hear back from me by end of week.  I took a quick look at the proposed fix and it looks promising.

**Comment by OmegaVVeapon** [comment]
@mattcarrier Hey, the week passed and we didn't get a response from you.

Did you get a chance to try the build to see if it solved your issue?

I'll leave this issue open until Friday but if there's no response by then, I'll assume everything worked fine and tag the release.

**Comment by OmegaVVeapon** [comment]
I have merged the PR and released [kopf-k8s-sidecar 1.1.1](https://github.com/OmegaVVeapon/kopf-k8s-sidecar/releases/tag/1.1.1).

The new Docker image has been published to [Dockerhub](https://hub.docker.com/r/omegavveapon/kopf-k8s-sidecar/tags)

Closing.
